Parquet{Read,Writ}er: block files when element count would be too large

tylershunt opened the pull request:
This is a work around for an apparent limitation in the arrow library
when translating from parquet

String arrays have a de-facto limit in the number of elements they can
contain (it seems close to the number of string bytes). You can end up
with more elements than string bytes if there are mostly null entries.

We can account for this in memory by chunking arrays, but parquet does
not have this limitation and the chunking is lost in translation. Then
when the arrays are translated back into arrow Sting Builder complains
that there are too many elements for the array.

The workaround here chunks parquet files as well.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

gurbinder533 commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -285,40 +282,40 @@ class Partitioner {
-        galois::runtime::gSerialize(bitsetBuffer, syncNodes[h]);
-        bytesSent += bitsetBuffer.size();
-        net.sendTagged(h, galois::runtime::evilPhase, bitsetBuffer);
+    galois::StatTimer timer("Phase0SendRecvBitsets", kGRName);
```
Remove phase0 from the timer name as well.

gurbinder533 commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -377,34 +374,34 @@ class Partitioner {
-    unsigned bytesSent = 0;
-    galois::StatTimer sendTimer("Phase0AsyncSendLoadTime", kGRName);
+    unsigned bytes_sent = 0;
+    galois::StatTimer timer("Phase0AsyncSendLoadTime", kGRName);
```
Same.

roshandathathri commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -181,43 +179,42 @@ class Partitioner {
-    bitsetSetupTimer.start();
+  void DoMasterAssignmentBitsetSetup(const PartialV1GraphView& pview,
+                                     DynamicBitSet& ghosts) {
+    galois::StatTimer timer("Phase0BitsetSetup", kGRName);
```
Function name as timer name

roshandathathri commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -229,53 +226,53 @@ class Partitioner {
+      DynamicBitSet& ghosts,
+      std::unordered_map<uint64_t, uint32_t>& gid_to_offsets,
+      galois::gstl::Vector<galois::gstl::Vector<uint32_t>>& sync_nodes) {
+    galois::StatTimer timer("Phase0MapSetup", kGRName);
```
Ditto

roshandathathri commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -377,34 +374,34 @@ class Partitioner {
-    sendTimer.stop();
+    timer.stop();
 
     galois::runtime::reportStat_Tsum(kGRName, "Phase0AsyncSendLoadBytesSent",
```
Ditto

roshandathathri commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -573,64 +571,62 @@ class Partitioner {
+    uint64_t bytes_sent = 0;
 
-    galois::StatTimer sendOffsetsTimer(statString.c_str(), kGRName);
+    galois::StatTimer timer("Phase0SendOffsets", kGRName);
```
Ditto

roshandathathri commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -573,64 +571,62 @@ class Partitioner {
 
-    galois::runtime::reportStat_Tsum(kGRName, statString + "BytesSent",
-                                     bytesSent);
+    galois::runtime::reportStat_Tsum(kGRName, "Phase0SendOffsetsBytesSent",
```
Ditto

roshandathathri commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -644,90 +640,90 @@ class Partitioner {
+      uint32_t begin, uint32_t end, uint32_t num_local_nodes,
+      std::vector<uint32_t>& local_node_to_master,
+      galois::gstl::Vector<galois::gstl::Vector<uint32_t>>& sync_nodes) {
+    galois::StatTimer timer("Phase0AssignmentSendTime", kGRName);
```
Ditto

roshandathathri commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -836,48 +834,45 @@ class Partitioner {
+  void RecvMasterAssignments(
+      std::vector<uint32_t>& local_node_to_master,
+      std::unordered_map<uint64_t, uint32_t>& gid_to_offsets) {
+    galois::StatTimer timer("Phase0AssignmentReceiveTime", kGRName);
```
Ditto

roshandathathri commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -836,48 +834,45 @@ class Partitioner {
+      std::vector<uint32_t>& local_node_to_master,
+      std::unordered_map<uint64_t, uint32_t>& gid_to_offsets,
+      DynamicBitSet& host_finished) {
+    galois::StatTimer timer("Phase0AssignmentReceiveTimeAsync", kGRName);
```
Ditto

roshandathathri commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -1335,22 +1297,22 @@ class Partitioner {
-   * Given a loaded graph, construct the edges in the DistGraph graph.
-   * Variant that constructs edge data as well.
+   * DistributeEdgeCut constructs edges assuming the graph is already partially
+   * constructed by DoEdgeAssignment.
```
the graph is already allocated and edge-indices is already constructed based on DoEdgeAssignment

Commit:
Parquet{Read,Writ}er: block files when element count would be too large

This is a work around for an apparent limitation in the arrow library
when translating from parquet

String arrays have a de-facto limit in the number of elements they can
contain (it seems close to the number of string bytes). You can end up
with more elements than string bytes if there are mostly null entries.

We can account for this in memory by chunking arrays, but parquet does
not have this limitation and the chunking is lost in translation. Then
when the arrays are translated back into arrow Sting Builder complains
that there are too many elements for the array.

The workaround here chunks parquet files as well.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

ddn0 commented on libtsuba/src/ParquetReader.cpp:
```diff
@@ -71,70 +72,41 @@ HandleBadParquetTypes(std::shared_ptr<arrow::ChunkedArray> old_array) {
+  KATANA_CHECKED_CONTEXT(
+      fv_tmp->Bind(
+          uri, 0, preload ? std::numeric_limits<uint64_t>::max() : 0, false),
+      uri);
```
`..., "opening {}", uri); `?

ddn0 commented on libtsuba/src/ParquetReader.cpp:
```diff
@@ -148,141 +120,251 @@ tsuba::ParquetReader::ReadFromUriSliced(const katana::Uri& uri) {
     }
-    if (idx >= schema.num_fields()) {
+
+    auto it = std::upper_bound(
```
Are row_offsets unique? If so, is it possible to use `lower_bound` and avoid `it - 1` below? Or equivalently, how do we know that `it` is not `begin`?

ddn0 commented on libtsuba/src/ParquetReader.cpp:
```diff
@@ -148,141 +120,251 @@ tsuba::ParquetReader::ReadFromUriSliced(const katana::Uri& uri) {
+        if (auto arr_it = read_arrays.find(idx); arr_it != read_arrays.end()) {
+          column = arr_it->second;
+        } else {
+          auto status = reader->ReadColumn(idx, &column);
```
I believe `KATANA_CHECKED` knows about arrow status as well, so you should be able to do: `KATANA_CHECKED(reader->ReadColumn(idx, &column));`

pyrito commented on libtsuba/src/ParquetReader.cpp:
```diff
@@ -71,70 +72,41 @@ HandleBadParquetTypes(std::shared_ptr<arrow::ChunkedArray> old_array) {
-    return res.error().WithContext("opening {}", uri);
-  }
+BuildReader(
+    const std::string& uri, bool preload,
```
nit: Why use string `uri` here instead of `katana::Uri`?

tylershunt replied:
Oh it's a private function, and a bunch of places I call it with strings from `fmt::format` so it seemed silly to convert back and forth from known well formed uris.

ddn0 commented on libtsuba/src/ParquetWriter.cpp:
```diff
@@ -208,53 +258,39 @@ katana::Result<void>
+  table.reset();
+
+  uint32_t table_count = 0;
+  std::string parts_descirption;
```
"parts_description" and is this used?

ddn0 commented on libtsuba/src/ParquetReader.cpp:
```diff
@@ -148,141 +120,251 @@ tsuba::ParquetReader::ReadFromUriSliced(const katana::Uri& uri) {
-  out = out->Slice(row_offset, slice.length);
+class BlockedParquetReader {
+public:
+  static Result<std::unique_ptr<BlockedParquetReader>> Make(
```
perhaps some documentation about the different file formats that this method supports reading (parquet or offsets)?

pyrito reviewed (approved):
LGTM
